internal/core/config/sources/vault/keys.go

```diff
@@ -0,0 +1,41 @@
+package vault
+
+import (
+	"errors"
+	"strings"
+)
+
+var errInvalidPathFormat = errors.New("invalid Vault path/key combination, expected format <path>[<key>]")
+
+// Allows key to be dot-delimited to traverse nested maps
+func traverseToKey(data map[string]interface{}, key string) interface{} {
+	parts := strings.Split(key, ".")
+
+	for i, part := range parts {
+		partVal := data[part]
+		if i == len(parts)-1 {
+			return partVal
+		}
+
+		var ok bool
+		data, ok = partVal.(map[string]interface{})
+		if !ok {
+			return nil
+		}
+	}
+	return nil
+}
+
+// The config path is of the form `/path/to/secret[keys.to.data]`
+func splitConfigPath(pathAndKey string) (string, string, error) {
+	if !strings.HasSuffix(pathAndKey, "]") {
+		return "", "", errInvalidPathFormat
+	}
+
+	parts := strings.SplitN(pathAndKey, "[", 2)
```

Comment:
No, in the secret data maps they could be valid characters since the secret data maps have string keys and they aren't any reserved chars.  Paths are more strict so it can be assumed there are no brackets in them.